### PR TITLE
Space - Nora

### PR DESCRIPTION
**Pull request by thenora**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails |  The model holds the database to communicate with the controller.
Describe in your own words what the Controller is doing in Rails | the controller contains the "business logic" about how the program works.
Describe in your own words what the View is doing in Rails |  The view tells us how to display the information to the browser.
Describe an edge-case controller test you wrote | an edge case for the update function was "will redirect to the root page if given an invalid id"
What is the purpose of using strong params? (i.e. the params method in the controller) | strong params let us setup once for our controller, then we don't need to specify each field individually. It saves time while keeping things secure.
How are Rails migrations related to Rails models? | I don't understand the question. Is this about databases?
Describe one area of Rails that are still unclear on | See previous question.

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  get '/tasks/new', to: 'tasks#new', as: 'new_task' #gets form for a new task
+  post '/tasks', to: 'tasks#create'
+
+  # routes on individual tasks
+  get '/tasks/:id', to: 'tasks#show', as: 'task' #shows details for one task
+  get '/tasks/:id/edit', to: 'tasks#edit', as: 'edit_task' #brings up edit form to edit one task
+  patch '/tasks/:id', to: 'tasks#update'
+  delete '/tasks/:id', to: 'tasks#destroy'
+
+  patch '/tasks/:id/mark_done', to: 'tasks#mark_done', as: 'mark_done'
```
Comment (beccaelenzil) [inline config/routes.rb:18]: Consider how making two separate methods, for instance `mark_complete` and `mark_incomplete` would make this action idempotent.

File: test/controllers/tasks_controller_test.rb [hunk 2]
```diff
@@ -88,33 +108,82 @@
...
   describe "destroy" do
     # Your tests go here
-    
+    it "will remove an item from the db and the task count will decrease" do
+      Task.create(name: "Get Shit Done", description: "You gotta werk.")
+
+      id = Task.last.id
+      expect {
+        delete task_path(id)
+      }.must_change "Task.count"
```
Comment (beccaelenzil) [inline test/controllers/tasks_controller_test.rb:184]: Consider an even more precise test.

```suggestion
      expect {
        delete task_path(id)
      }.will_differ "Task.count", -1
```

File: test/controllers/tasks_controller_test.rb [hunk 3]
```diff
@@ -88,33 +108,82 @@
...
+
```
Comment (beccaelenzil) [inline test/controllers/tasks_controller_test.rb:185]: Remember to test for a redirect

File: test/controllers/tasks_controller_test.rb [hunk 4]
```diff
@@ -88,33 +108,82 @@
...
   
   # Uncomment and complete these tests for Wave 3
   describe "update" do
+    
     # Note:  If there was a way to fail to save the changes to a task, that would be a great
     #        thing to test.
+
     it "can update an existing task" do
-      # Your code here
+      Task.create(name: "Get Shit Done", description: "You gotta werk.")
```
Comment (beccaelenzil) [inline test/controllers/tasks_controller_test.rb:140]: Good work fleshing out this test.

File: app/views/tasks/_form.html.erb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+<%= form_with model: @task, class: 'create-form' do |f| %>
```
Comment (beccaelenzil) [inline app/views/tasks/_form.html.erb:1]: Nice use of partial views.

File: app/views/tasks/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,3 @@
+<h2>Add a Task</h2>
```
Comment (beccaelenzil) [inline app/views/tasks/new.html.erb:1]: Consider adding this to the partial by adding to the locals hash.

**Comment by beccaelenzil** [comment]
# Task List




## Major Learning Goals/Code Review





| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
| At least 6 commits with meaningful commit messages | ✔️
| Routes follow RESTful conventions | ✔️
| Uses named routes (like `_path`) | ✔️
| Creates Models and migrations | ✔️
| Creates styled views | No, but not required
| Handles errors like nonexistant tasks | ✔️
| Uses `form_with` to render forms in Rails | ✔️

## Functional Requirements/Manual Testing

| Functional Requirement | yes/no |
| --- | --- |
| Successfully handles index & show | ✔️
| index & show tests pass | ✔️
| Successfully handles: New, Create | ✔️
| New, Create tests pass | ✔️
| Successfully handles: Edit, Update | ✔️
| Successfully handles: Destroy, Task Complete | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 5+ in Code Review && 5+ in Functional Requirements | ✔️






## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅

**Review by beccaelenzil — COMMENTED** [review]
Great work on this project! It's clear that the learning goals for your first rails project were met. Keep up the hard work.
